Re-PR add stm32h755II chip

jfbblue0922 opened the pull request:
@acassis
@davids5

we made PR again.

・Add new CONFIG_STM32H7_PWR_LDO_SUPPLY macro to "Kconfig".
・Delete "stm32h7x5xx_rcc.c" and add ONFIG_STM32H7_PWR_LDO_SUPPLY macro to "stm32h7x3xx_rcc.c".

davids5 commented on arch/arm/src/stm32h7/Kconfig:
```diff
@@ -284,11 +284,24 @@ config STM32H7_HAVE_SMPS
 
+choice
+	prompt "STM32 H7 Power Supply Selection"
+	default STM32H7_PWR_DIRECT_SMPS_SUPPLY
```
I do not think the default is correct.  Before the change the what branch of the https://github.com/apache/nuttx/pull/12394/files#diff-9dfe671931738d39252fc8e902e892d172ac118551d14e1a454e324edb0d0be4R889-R897 would have been selected.

We have to be super carful here not break existing boards. That had the default set BEFORE this change

jfbblue0922 replied:
I changed default value.

davids5 commented on arch/arm/src/stm32h7/Kconfig:
```diff
@@ -284,11 +284,27 @@ config STM32H7_HAVE_SMPS
+
+config STM32H7_PWR_DEFAULT_SUPPLY
+	bool "Default"
+
```
Add help to all these  `config`s to describe what mode it is, based on the bits selected by `STM32 H7 Power Supply Selection`

jfbblue0922 replied:
@davids5
We tried putting other modes into Kconfig.

jfbblue0922 commented on arch/arm/src/stm32h7/stm32h7x3xx_rcc.c:
```diff
@@ -886,6 +886,12 @@ void stm32_stdclockconfig(void)
+          STM32_PWR_CR3_SMPSEXTHP | STM32_PWR_CR3_SMPSLEVEL_MASK);
+      regval |= STM32_PWR_CR3_LDOEN;
+      putreg32(regval, STM32_PWR_CR3);
 #else
```
I added comment.

davids5 commented on arch/arm/src/stm32h7/stm32h7x3xx_rcc.c:
```diff
@@ -886,7 +886,19 @@ void stm32_stdclockconfig(void)
+      regval |= STM32_PWR_CR3_LDOEN;
+      putreg32(regval, STM32_PWR_CR3);
 #else
+      /* 4 else cases of power supply mode
```
There are 3 settings. 2 are covered above.

What is the "4 else cases of power supply mode" referring too. This it the third. It is the default STM32H7_PWR_DEFAULT_SUPPLY.  should there be another else with #error "No a STM32 H7 Power Supply Selection made"?

jfbblue0922 replied:
@davids5
STM32H7 has six power supply modes to choose from.
We use LDO supply.
I suspect someone is using a DIRECT SMPS supply.
And I don't know what the first person used.
So we don't know what the 'else' power mode is.

'else' is considered to be when nothing is defined, but I don't know how to change 'default' from 'defining nothing' in 'Kconfig'.

What do you think we should do?

davids5 reviewed (changes requested):
Please add help and comments as suggested and see if the #error make sense to add.

davids5 commented on arch/arm/src/stm32h7/Kconfig:
```diff
@@ -284,11 +284,55 @@ config STM32H7_HAVE_SMPS
-config STM32H7_PWR_DIRECT_SMPS_SUPPLY
-	bool "Use direct SMPS supply mode"
+choice
+	prompt "STM32 H7 Power Supply Selection"
```
@jfbblue0922 Please build test these changes for other version of the IC. 

STM32H7_PWR_DEFAULT_SUPPLY - is not defined
STM32H7_HAVE_SMPS - will disable the choices

jfbblue0922 replied:
I removed the 'STM32H7_PWR_DEFAULT_SUPPLY' macro in Kconfig.

I tried setting the 'STM32H7_HAVE_SMPS' macro in Kconfig to other chips.
- The "STM32 H7 Power Supply Selection" menu does not appear for STM32H7x3xx
- The "STM32 H7 Power Supply Selection" menu appears for STM32H7x5xx or STM32H7x7xx

davids5 commented on arch/arm/src/stm32h7/Kconfig:
```diff
@@ -284,11 +284,55 @@ config STM32H7_HAVE_SMPS
+	    Note:The SMPS step-down converter is not available on all packages.
+
+        Select the power supply mode of your board
+	    - ID:1 LDO supply mode used 'STM32H7_PWR_LDO_SUPPLY' macro
```
How does ID:x relate to the choices?

jfbblue0922 replied:
This ID is from the data sheet (Reference manual RM0399).
Power control (PWR)
- Figure 22. System supply configurations
![Figure22](https://github.com/apache/nuttx/assets/113963003/18d36af0-652f-4a6b-b032-d0970c4ace03)

- Table 33. Supply configuration control
![Table33](https://github.com/apache/nuttx/assets/113963003/aa00822a-cdfd-4f3b-9601-3c4ac416c8a7)

davids5 commented on arch/arm/src/stm32h7/stm32h7x3xx_rcc.c:
```diff
@@ -886,6 +886,20 @@ void stm32_stdclockconfig(void)
+      /* These 4 cases of power supply mode have no macros created.
+       * Selecting one of these modes will cause a compile error.
+       */
+      #error Macro of this power supply mode is not created yet.
```
What would a user be expected to do if they get this message?

jfbblue0922 replied:
Since I cannot verify the operation of these modes, I hope that those who use them will add macros.

davids5 replied:
The correct thing to do is: The Kconfig should name the 6 configurations available based on the package and IC Family  and the code should set them. The default should be what it was before this PR. 

Putting a partial solution and jeopardizing breaking existing boards is totally un-acceptable.

jfbblue0922 replied:
@davids5
Before this PR, we could only use DIRECT_SMPS_SYPPLY mode or else.
And we just want to use LDO_SUPPLY mode.
Should we remove the 4 modes that no one is currently using?

And we are trying to be careful not to break the existing board.

davids5 replied:
@jfbblue0922 I will spend some time in the next couple of days reviewing the data sheets the  and push some changes to this PR.

davids5 reviewed (changes requested):
The correct thing to do is: The Kconfig should name the 6 configurations available based on the package and IC Family  and the code should set them. The default should be what it was before this PR. 

Putting a partial solution and jeopardizing breaking existing boards is totally un-acceptable.

davids5 commented:
@jfbblue0922 Please see https://github.com/nuttx-to-asf/incubator-nuttx/commit/c6cedbaa23161950b18caed94af54b549ef8a4b1 
It is the approach I had suggested originally. It adds the LDO mode with a chip the has SMPS. Please cherry-pick it to your branch and test it does what you needed. If it is fine then please squash it and force push.

jfbblue0922 commented:
@davids5

I agree with your method.
I've confirmed that this method works correctly.
Thank you for your help.

davids5 commented:
> @davids5
> 
> I agree with your method. I've confirmed that this method works correctly. Thank you for your help.

Great! You are very welcome. Now we just have to get CI to be happy.

Please rebase on current NuttX master and force push.

davids5 commented:
@jfbblue0922 - hmm Look Like CI is still not happy. @xiaoxiang781216 Any ideas what it broken?

xiaoxiang781216 commented:
ci break is fixed one week ago, @jfbblue0922  please rebase your patch to the latest master.

jfbblue0922 commented:
@davids5
@xiaoxiang781216

I rebased and pushed but nothing changed.
What should we do?

xiaoxiang781216 commented on arch/arm/src/stm32h7/Kconfig:
```diff
@@ -284,11 +284,23 @@ config STM32H7_HAVE_SMPS
-	bool "Use direct SMPS supply mode"
+choice
+	prompt "STM32 H7 Power Supply Selection"
+	default STM32H7_PWR_DIRECT_SMPS_SUPPLY
```
where you define STM32H7_PWR_DIRECT_SMPS_SUPPLY

xiaoxiang781216 commented on arch/arm/src/stm32h7/Kconfig:
```diff
@@ -284,11 +284,23 @@ config STM32H7_HAVE_SMPS
+	bool "Use LDO supply mode"
 	default n
 
+endchoice #"STM32 H7 Power Supply Selection"
```
```suggestion
endchoice

jfbblue0922 replied:
@xiaoxiang781216 

I corrected it.

davids5 commented on arch/arm/src/stm32h7/Kconfig:
```diff
@@ -284,10 +284,24 @@ config STM32H7_HAVE_SMPS
+choice
+	prompt "STM32 H7 Power Supply Selection"
+	default STM32H7_PWR_DIRECT_SMPS_SUPPLY
+	depends on !ARCH_BOARD_NUCLEO_H745ZI && !NUCLEOH745ZI_SMPS
```
A board can not be used here.

jfbblue0922 replied:
@acassis 

Thank you for your advice.

xiaoxiang781216 commented on boards/Kconfig:
```diff
@@ -1747,6 +1747,7 @@ config ARCH_BOARD_NUCLEO_H745ZI
 	depends on ARCH_CHIP_STM32H745ZI
 	select ARCH_HAVE_LEDS
 	select ARCH_HAVE_BUTTONS
+	select HAS_PWR_DIRECT_SMPS_SUPPLY
```
where you define HAS_PWR_DIRECT_SMPS_SUPPLY

jfbblue0922 replied:
@xiaoxiang781216 

I added HAS_PWR_DIRECT_SMPS_SUPPLY to stm32h7/Kconfig.
Thank you for your advice.

xiaoxiang781216 commented on arch/arm/src/stm32h7/Kconfig:
```diff
@@ -284,10 +284,24 @@ config STM32H7_HAVE_SMPS
+choice
+	prompt "STM32 H7 Power Supply Selection"
+	default STM32H7_PWR_DIRECT_SMPS_SUPPLY
+	depends on STM32H7_HAVE_SMPS && !HAS_PWR_DIRECT_SMPS_SUPPLY
```
where you define HAS_PWR_DIRECT_SMPS_SUPPLY

jfbblue0922 replied:
@xiaoxiang781216 

Is HAS_PWR_DIRECT_SMPS_SUPPLY defined in the wrong place?
I'm not sure where to define it correctly.
If it's in the wrong place, please tell me the correct place.

xiaoxiang781216 replied:
All option must be defined through config keyword before using, but where do you define the new option HAS_PWR_DIRECT_SMPS_SUPPLY?

jfbblue0922 replied:
I corrected it.

jfbblue0922 commented on arch/arm/src/stm32h7/hardware/stm32h7x3xx_pwr.h:
```diff
@@ -129,16 +129,18 @@
-#  define STM32_PWR_CR3_SMPSLEVEL_1V8       (1 << STM32_PWR_CR3_SMPSLEVEL_SHIFT) /* 01 */
-#  define STM32_PWR_CR3_SMPSLEVEL_2V5       (2 << STM32_PWR_CR3_SMPSLEVEL_SHIFT) /* 10: */
-#  define STM32_PWR_CR3_SMPSLEVEL_2V5B      (3 << STM32_PWR_CR3_SMPSLEVEL_SHIFT) /* 11: */
+#  define STM32_PWR_CR3_SDEN                  (1 << 2)  /* Bit 2: SMPS step-down converter enable */
```
I corrected it.

jfbblue0922 replied:
@xiaoxiang781216 

I've removed the spaces and aligned the values of define.

xiaoxiang781216 commented on arch/arm/src/stm32h7/Kconfig:
```diff
@@ -284,11 +284,34 @@ config STM32H7_HAVE_SMPS
 	bool
 	default n
 
+config HAS_PWR_DIRECT_SMPS_SUPPLY
```
no change

jfbblue0922 replied:
@xiaoxiang781216 

I changed it from HAS_PWR_DIRECT_SMPS_SUPPLY to STM32H7_HAVE_PWR_DIRECT_SMPS_SUPPLY.

jerpelea reviewed (commented):
please add commit message

jfbblue0922 commented:
@jerpelea 

I changed it from HAS_PWR_DIRECT_SMPS_SUPPLY to STM32H7_HAVE_PWR_DIRECT_SMPS_SUPPLY
and I rebased onto the current NuttX master and force pushed.

jerpelea commented:
> @jerpelea
> 
> I changed it from HAS_PWR_DIRECT_SMPS_SUPPLY to STM32H7_HAVE_PWR_DIRECT_SMPS_SUPPLY and I rebased onto the current NuttX master and force pushed.

if you check the patch there is no commit message
you can reuse the same message as the one it PR
git commit --amend 
type the message and save
git push -f 
this will update the current patch without the need to abort it 

Best regards
Alin
